### PR TITLE
Add enum support for DCAs and models

### LINKED ISSUE #6546: First class enum support in DCA option fields
### Description

Since the introduction of enums in PHP, their advantage has become more and more obvious and we all love the added expressiveness and the link between values and their associated logic

It would therefore be great if enums were also supported as options in DCA fields. This may also open up the possibility for the core itself to use enums for particular configuration options.

From my point of view, support can be broken down into three parts:
1. defining an enum in a DCA field and using the enum cases as options for the field
2. automatic translation of the enum values via Symfony Translation
3. bonus: an option to resolve an enum from its value within a Contao model

Implementation might be possible as follows:

### 1. Generation of Options

To minimize the need for changes, it would be sufficient to fill the existing `options` array from the enum cases.

The prerequisite for this is that the linked enum is a `BackedEnum`.

Example:
````php
# Within a DCA field
[
  'inputType' => 'select',
  // Must implement \BackedEnum
  'enum' => App\Data\Salutation::class,
]
````

The implementation would be possible via a `loadDataContainer` hook.

### 2. Automatic Translation of Values

For convenient use in the backend, matching translations should be generated in place of the enum values for the options.

For this purpose, an interface could be used as follows:

````php
use Symfony\Component\Translation\TranslatableMessage;

interface LabeledEnum
{
    public function label(): TranslatableMessage;
}
````

Within the DCA hook, translations can then be automatically generated for the enum and stored as `references` in the DCA field.

The big advantage compared to a manual assignment of `references` is the possibility to translate the enum uniformly and especially outside of the DCA / outside of Contao.

Note: Contao internal translations may still be used by delegating to the Contao domain within the `TranslatableMessage`. In addition, the manual provision of `references` for enum options also remains possible.

### 3. Resolving the Enum in Models

In addition to the use within the backend / within a DCA, the resolving of an enum within a model  provides another comfort to developers.

Imagine a model method as follows:

````php
public function getEnum(string $key): ?BackedEnum
````

Which in turn can be used by developers in their models:

````php
public function getSalutation(): Salutation
{
    return $this->getEnum('salutation') ?? Salutation::Generic:
}
````

As a model itself has no access to the DCA, an extension to the `DcaExtractor` would probably be required for the implementation.

----

I would like to hear feedback on whether any of the three parts are interesting for the core, and if so, which ones.

Since we use enums pretty much as described above, I may take care of the implementation.

### PR DESCRIPTION
**Pull request by SeverinGloeckle**
Implements #6546 

A few notes on the implementation:

### 1. Usage of TranslatableMessage

The generation of translated references for the generated options is implemented via Symfony’s `TranslatableMessage`.

In my opinion, this is the most sane way to handle translations in a complex system with various sources of translation.

Using Symfony’s `TranslatableInterface` however is less restrictive, but does not provide the additional information on parameters and especially on the translation domain.

Although it is outside the scope of this PR, a forward-looking decision ought to be made on how Contao wants to deal with translations in the future. Up to now, neither of the two interfaces is being used.

### 2. Validation of enum values in the model

It’s not straightforward to define a proper handling of invalid enum values in the model. One could:
* simply return null for invalid values
* raise an exception explaining the invalid value (like implemented)
  * Since there is no built-in way to detect whether an enum is backed by strings or integers, only a generic exception for both types is possible (without adding somethin like an `EnumUtil`).
* let the call to `BackedEnum::tryFrom()` fail with the invalid value and let it throw its TypeError

I implemented the second way in order to give developers a proper hint pointing to the DCA configuration but the other two ways are reasonable as well.

### 3. Modifications & unit tests for legacy code

For the changes to the legacy code, I stuck to existing styles and structures as much as possible.

For the model, only a simple unit test is added for resolving the enum - if necessary, I can also add tests for the various exceptions.

### 4. Needs documentation

If this will be merged, I'll also add corresponding notes to the developer documentation.

### CODE REVIEW

**Comment by aschempp** [comment]
Any particular reason we need the DcaExtractor to read the enums? Why not just use the listener to convert enum to string options?

**Comment by SeverinGloeckle** [comment]
The DcaExtractor maps the individual fields to the configured enum classes and transfers this assignment to the model.

To resolve the enums via `Model::getEnum()`, the specific enum class is required in addition to the value of the field (see example in third section of #6546)

**Review by leofeyer — DISMISSED** [review]
I can confirm that the code works. 👍

**Comment by leofeyer** [comment]
@aschempp Has your question above been answered?

@contao/developers Shall we add two example enums so people can use them as templates for their own enums?

```php
namespace Contao\CoreBundle\DataContainer\Enum;

enum PageRobotsEnum: string
{
    case indexFollow = 'index,follow';
    case indexNofollow = 'index,nofollow';
    case noindexFollow = 'noindex,follow';
    case noindexNofollow = 'noindex,nofollow';
}
```

```php
namespace Contao\CoreBundle\DataContainer\Enum;

use Contao\CoreBundle\Translation\TranslatableLabelInterface;
use Symfony\Component\Translation\TranslatableMessage;

enum PageRedirectEnum: string implements TranslatableLabelInterface
{
    case permanent = 'permanent';
    case temporary = 'temporary';

    public function label(): TranslatableMessage
    {
        return new TranslatableMessage('tl_page.'.$this->name, [], 'contao_tl_page');
    }
}
```

Do you agree with the namespace?

File: core-bundle/tests/Fixtures/Enum/IntBackedEnum.php [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+ *
+ * @license LGPL-3.0-or-later
+ */
+
+namespace Contao\CoreBundle\Tests\Fixtures\Enum;
+
+enum IntBackedEnum: int
+{
+    case OptionA = 13;
+    case OptionB = 42;
```
Comment (ausi) [inline core-bundle/tests/Fixtures/Enum/IntBackedEnum.php]: Just a nitpick, but our namig convention for enum cases is lowerCamelCase, see for example: https://github.com/contao/contao/blob/fd38b7fc2de60b418b835400087f42a3dc9de10c/core-bundle/src/Filesystem/SortMode.php

```suggestion
    case optionA = 13;
    case optionB = 42;
```

**Comment by Toflar** [comment]
@SeverinGloeckle any reason you did not go for `'options' -> Enum` but a new key instead?

File: core-bundle/tests/Fixtures/Enum/IntBackedEnum.php (hunk 1 above)
Comment (SeverinGloeckle) [reply core-bundle/tests/Fixtures/Enum/IntBackedEnum.php]: Both enum fixtures and the test have been adjusted in https://github.com/contao/contao/pull/6584/commits/9b722df4d63e0628256195ef041f461222c5cb5b

**Comment by SeverinGloeckle** [comment]
> @contao/developers Shall we add two example enums so people can use them as templates for their own enums?
> 
> ```
> enum PageRobotsEnum: string
> ```
> 
> ```
> enum PageRedirectEnum: string implements TranslatableLabelInterface
> ```

The two examples are a very good idea. However, I would prefer to add them in a follow-up PR for an independent discussion  (namespace, extensibility, possibly resolution within the `PageModel`).

I may create a follow-up PR as soon as this PR has been merged.

**Comment by SeverinGloeckle** [comment]
> @SeverinGloeckle any reason you did not go for `'options' -> Enum` but a new key instead?

From my point of view, the explicit tagging of the enum provides the following benefits:
1. unambiguous check whether a DCA field is linked to an enum (no type check)
2. generation of `references` even if the `options` were overwritten manually (see 90f9e22dc885c617d9ca9fbd482f8088d9231ee9)
3. no side effects on third party code (which may rely on `options` being an array)

**Review by Toflar — APPROVED** [review]
I'm fine with that

**Comment by leofeyer** [comment]
Thank you @SeverinGloeckle.

**Pull request merged**